### PR TITLE
yarn-berry: init at 3.4.1

### PR DESCRIPTION
**Pull request by ryota-ka**
###### Description of changes

Added the newer implementation (a.k.a. "Berry") of [Yarn](https://yarnpkg.com/) package manager.

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- [x] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/tools/yarn-berry/default.nix [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+    nodejs
+  ];
+
+  dontConfigure = true;
+
+  dontBuild = true;
+
+  installPhase = ''
+    runHook preInstall
+    install -Dm 755 "$src/packages/yarnpkg-cli/bin/yarn.js" "$out/bin/yarn"
```
Comment (dotlambda) [inline pkgs/development/tools/yarn-berry/default.nix:24]: Can we build this file manually?

File: pkgs/development/tools/yarn-berry/default.nix [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+stdenv.mkDerivation rec {
+  name = "yarn-berry";
+  version = "3.4.1";
+
+  src = fetchFromGitHub {
+    owner = "yarnpkg";
+    repo = "berry";
+    rev = "@yarnpkg/cli/${version}";
+    sha256 = "sha256-eBBB/F+mnGi93Qf23xgt306/ogoV76RXOM90O14u5Tw=";
```
Comment (dotlambda) [inline pkgs/development/tools/yarn-berry/default.nix:11]:
```suggestion
    hash = "sha256-eBBB/F+mnGi93Qf23xgt306/ogoV76RXOM90O14u5Tw=";
```

File: pkgs/development/tools/yarn-berry/default.nix (hunk 1 above)
Comment (ryota-ka) [reply pkgs/development/tools/yarn-berry/default.nix:24]: Okay, I'll try it.

Comment (ryota-ka) [reply pkgs/development/tools/yarn-berry/default.nix]: @dotlambda I've pushed a commit so that we build it manually.
Could you please take another look?

**Comment by emptyflask** [comment]
I tested this on nixos / x86_64-linux with a real project and it seems to work

File: pkgs/development/tools/yarn-berry/default.nix [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+  src = fetchFromGitHub {
+    owner = "yarnpkg";
+    repo = "berry";
+    rev = "@yarnpkg/cli/${version}";
+    hash = "sha256-eBBB/F+mnGi93Qf23xgt306/ogoV76RXOM90O14u5Tw=";
+  };
+
+  buildInputs = [
+    nodejs
+    yarn
```
Comment (SuperSandro2000) [inline pkgs/development/tools/yarn-berry/default.nix:16]: yarn should go to nativeBuildInputs

Comment (ryota-ka) [reply pkgs/development/tools/yarn-berry/default.nix:19]: Fixed in 05f304499fdc973573c3cff68ee20ac659e2d46c
Could you take another look please?

**Review by SuperSandro2000 — APPROVED** [review]

**Comment by ryota-ka** [comment]
@SuperSandro2000 Thank you for your review and rebasing the PR!

**Pull request merged**